File: apk-permission-manager/axml.js
/*
 * Minimal binary AndroidManifest.xml (AXML) reader/writer.
 * Implements just enough of the AOSP ResourceTypes.h chunk format to
 * round-trip a manifest: string pool, optional resource map, and the
 * flat namespace/element/cdata node stream. Runs in the browser and in
 * Node (no DOM / browser globals used), so the same file is used for
 * both the app and the test harness.
 */
(function (global) {
  'use strict';

  var CHUNK_STRING_POOL = 0x0001;
  var CHUNK_XML = 0x0003;
  var CHUNK_XML_START_NAMESPACE = 0x0100;
  var CHUNK_XML_END_NAMESPACE = 0x0101;
  var CHUNK_XML_START_ELEMENT = 0x0102;
  var CHUNK_XML_END_ELEMENT = 0x0103;
  var CHUNK_XML_CDATA = 0x0104;
  var CHUNK_XML_RESOURCE_MAP = 0x0180;

  var NO_INDEX = 0xFFFFFFFF;

  function decodeUtf16LEString(bytes, offset, charLen) {
    // Decode charLen UTF-16 code units starting at byte `offset`.
    var units = new Uint16Array(charLen);
    for (var i = 0; i < charLen; i++) {
      units[i] = bytes[offset + i * 2] | (bytes[offset + i * 2 + 1] << 8);
    }
    var s = '';
    var CHUNK = 8192;
    for (var i = 0; i < units.length; i += CHUNK) {
      s += String.fromCharCode.apply(null, units.subarray(i, i + CHUNK));
    }
    return s;
  }

  function readLen16(dv, pos) {
    var u0 = dv.getUint16(pos, true);
    if (u0 & 0x8000) {
      var u1 = dv.getUint16(pos + 2, true);
      return { value: ((u0 & 0x7FFF) << 16) | u1, size: 4 };
    }
    return { value: u0, size: 2 };
  }

  function readLen8(bytes, pos) {
    var b0 = bytes[pos];
    if (b0 & 0x80) {
      var b1 = bytes[pos + 1];
      return { value: ((b0 & 0x7F) << 8) | b1, size: 2 };
    }
    return { value: b0, size: 1 };
  }

  function parseStringPool(bytes, dv, base) {
    var chunkSize = dv.getUint32(base + 4, true);
    var stringCount = dv.getUint32(base + 8, true);
    var styleCount = dv.getUint32(base + 12, true);
    var flags = dv.getUint32(base + 16, true);
    var stringsStart = dv.getUint32(base + 20, true);
    var utf8 = (flags & 0x100) !== 0;

    var offsets = new Array(stringCount);
    for (var i = 0; i < stringCount; i++) {
      offsets[i] = dv.getUint32(base + 28 + i * 4, true);
    }

    var strings = new Array(stringCount);
    for (var i = 0; i < stringCount; i++) {
      var pos = base + stringsStart + offsets[i];
      if (utf8) {
        var l1 = readLen8(bytes, pos); // char length (unused)
        var l2 = readLen8(bytes, pos + l1.size); // byte length
        var dataPos = pos + l1.size + l2.size;
        var slice = bytes.subarray(dataPos, dataPos + l2.value);
        strings[i] = Buffer && Buffer.from ? Buffer.from(slice).toString('utf8') : utf8DecodeFallback(slice);
      } else {
        var lr = readLen16(dv, pos);
        strings[i] = decodeUtf16LEString(bytes, pos + lr.size, lr.value);
      }
    }

    return { chunkSize: chunkSize, strings: strings, styleCount: styleCount };
  }

  function utf8DecodeFallback(bytes) {
    if (typeof TextDecoder !== 'undefined') {
      return new TextDecoder('utf-8').decode(bytes);
    }
    var s = '';
    for (var i = 0; i < bytes.length; i++) s += String.fromCharCode(bytes[i]);
    return decodeURIComponent(escape(s));
  }

  function parse(buffer) {
    var bytes = buffer instanceof Uint8Array ? buffer : new Uint8Array(buffer);
    var dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);

    var rootType = dv.getUint16(0, true);
    if (rootType !== CHUNK_XML) {
      throw new Error('Not a binary XML document (unexpected root chunk type 0x' + rootType.toString(16) + ')');
    }
    var rootSize = dv.getUint32(4, true);

    var pos = 8; // past root chunk header
    if (dv.getUint16(pos, true) !== CHUNK_STRING_POOL) {
      throw new Error('Expected string pool chunk at offset ' + pos);
    }
    var pool = parseStringPool(bytes, dv, pos);
    var strings = pool.strings;
    pos += pool.chunkSize;

    var resourceMapIds = null;
    if (pos + 8 <= bytes.length && dv.getUint16(pos, true) === CHUNK_XML_RESOURCE_MAP) {
      var rmSize = dv.getUint32(pos + 4, true);
      var n = (rmSize - 8) / 4;
      resourceMapIds = new Array(n);
      for (var i = 0; i < n; i++) {
        resourceMapIds[i] = dv.getUint32(pos + 8 + i * 4, true);
      }
      pos += rmSize;
    }

    var nodes = [];
    var end = rootSize;
    while (pos < end) {
      var type = dv.getUint16(pos, true);
      var headerSize = dv.getUint16(pos + 2, true);
      var size = dv.getUint32(pos + 4, true);
      var lineNumber = dv.getUint32(pos + 8, true);
      var comment = dv.getUint32(pos + 12, true);

      if (type === CHUNK_XML_START_NAMESPACE || type === CHUNK_XML_END_NAMESPACE) {
        var prefix = dv.getUint32(pos + 16, true);
        var uri = dv.getUint32(pos + 20, true);
        nodes.push({
          type: type === CHUNK_XML_START_NAMESPACE ? 'nsStart' : 'nsEnd',
          lineNumber: lineNumber, comment: comment, prefix: prefix, uri: uri
        });
      } else if (type === CHUNK_XML_START_ELEMENT) {
        // attrExt fields sit right after the 16-byte node header, regardless
        // of the chunk's headerSize (which stays 16 for this chunk type too).
        var ns = dv.getUint32(pos + 16, true);
        var name = dv.getUint32(pos + 20, true);
        var attributeStart = dv.getUint16(pos + 24, true);
        var attributeSize = dv.getUint16(pos + 26, true);
        var attributeCount = dv.getUint16(pos + 28, true);
        var idIndex = dv.getUint16(pos + 30, true);
        var classIndex = dv.getUint16(pos + 32, true);
        var styleIndex = dv.getUint16(pos + 34, true);
        var attrArrayBase = pos + 16 + attributeStart;
        var attributes = [];
        for (var a = 0; a < attributeCount; a++) {
          var ab = attrArrayBase + a * attributeSize;
          attributes.push({
            ns: dv.getUint32(ab, true),
            name: dv.getUint32(ab + 4, true),
            rawValue: dv.getUint32(ab + 8, true),
            valueSize: dv.getUint16(ab + 12, true),
            valueRes0: dv.getUint8(ab + 14),
            valueType: dv.getUint8(ab + 15),
            valueData: dv.getUint32(ab + 16, true)
          });
        }
        nodes.push({
          type: 'elemStart', lineNumber: lineNumber, comment: comment,
          ns: ns, name: name, idIndex: idIndex, classIndex: classIndex, styleIndex: styleIndex,
          attributes: attributes
        });
      } else if (type === CHUNK_XML_END_ELEMENT) {
        var ns2 = dv.getUint32(pos + 16, true);
        var name2 = dv.getUint32(pos + 20, true);
        nodes.push({ type: 'elemEnd', lineNumber: lineNumber, comment: comment, ns: ns2, name: name2 });
      } else if (type === CHUNK_XML_CDATA) {
        var data = dv.getUint32(pos + 16, true);
        var vSize = dv.getUint16(pos + 20, true);
        var vRes0 = dv.getUint8(pos + 22);
        var vType = dv.getUint8(pos + 23);
        var vData = dv.getUint32(pos + 24, true);
        nodes.push({
          type: 'cdata', lineNumber: lineNumber, comment: comment, data: data,
          valueSize: vSize, valueRes0: vRes0, valueType: vType, valueData: vData
        });
      } else {
        // Unknown chunk type: keep raw bytes so we can round-trip untouched.
        nodes.push({ type: 'raw', bytes: bytes.slice(pos, pos + size) });
      }
      pos += size;
    }

    return { strings: strings, resourceMapIds: resourceMapIds, nodes: nodes };
  }

  // ---- serialization ----

  function utf8Encode(str) {
    if (typeof TextEncoder !== 'undefined') return new TextEncoder().encode(str);
    return new Uint8Array(Buffer.from(str, 'utf8'));
  }

  function buildStringPool(strings) {
    var stringCount = strings.length;
    var offsetsSize = stringCount * 4;
    var stringsStart = 28 + offsetsSize;

    var encoded = [];
    var dataLen = 0;
    for (var i = 0; i < stringCount; i++) {
      var s = strings[i];
      var len = s.length; // UTF-16 code unit count == JS string length
      var headBytes;
      if (len >= 0x8000) {
        headBytes = 4;
      } else {
        headBytes = 2;
      }
      var byteLen = headBytes + len * 2 + 2; // + null terminator
      encoded.push({ str: s, len: len, headBytes: headBytes, byteLen: byteLen });
      dataLen += byteLen;
    }
    var paddedDataLen = dataLen;
    while ((stringsStart + paddedDataLen) % 4 !== 0) paddedDataLen++;

    var chunkSize = stringsStart + paddedDataLen;
    var out = new Uint8Array(chunkSize);
    var dv = new DataView(out.buffer);

    dv.setUint16(0, CHUNK_STRING_POOL, true);
    dv.setUint16(2, 28, true);
    dv.setUint32(4, chunkSize, true);
    dv.setUint32(8, stringCount, true);
    dv.setUint32(12, 0, true); // styleCount
    dv.setUint32(16, 0, true); // flags: UTF-16, not sorted
    dv.setUint32(20, stringsStart, true);
    dv.setUint32(24, 0, true); // stylesStart

    var offset = 0;
    var dataPos = stringsStart;
    for (var i = 0; i < stringCount; i++) {
      dv.setUint32(28 + i * 4, offset, true);
      var e = encoded[i];
      var p = dataPos + offset;
      if (e.headBytes === 4) {
        dv.setUint16(p, 0x8000 | (e.len >>> 16), true);
        dv.setUint16(p + 2, e.len & 0xFFFF, true);
      } else {
        dv.setUint16(p, e.len, true);
      }
      var charPos = p + e.headBytes;
      for (var c = 0; c < e.len; c++) {
        dv.setUint16(charPos + c * 2, e.str.charCodeAt(c), true);
      }
      dv.setUint16(charPos + e.len * 2, 0, true); // terminator
      offset += e.byteLen;
    }

    return out;
  }

  function buildResourceMap(ids) {
    var chunkSize = 8 + ids.length * 4;
    var out = new Uint8Array(chunkSize);
    var dv = new DataView(out.buffer);
    dv.setUint16(0, CHUNK_XML_RESOURCE_MAP, true);
    dv.setUint16(2, 8, true);
    dv.setUint32(4, chunkSize, true);
    for (var i = 0; i < ids.length; i++) dv.setUint32(8 + i * 4, ids[i], true);
    return out;
  }

  function buildNode(node) {
    var out, dv;
    switch (node.type) {
      case 'nsStart':
      case 'nsEnd':
        out = new Uint8Array(24);
        dv = new DataView(out.buffer);
        dv.setUint16(0, node.type === 'nsStart' ? CHUNK_XML_START_NAMESPACE : CHUNK_XML_END_NAMESPACE, true);
        dv.setUint16(2, 16, true);
        dv.setUint32(4, 24, true);
        dv.setUint32(8, node.lineNumber >>> 0, true);
        dv.setUint32(12, node.comment >>> 0, true);
        dv.setUint32(16, node.prefix >>> 0, true);
        dv.setUint32(20, node.uri >>> 0, true);
        return out;

      case 'elemEnd':
        out = new Uint8Array(24);
        dv = new DataView(out.buffer);
        dv.setUint16(0, CHUNK_XML_END_ELEMENT, true);
        dv.setUint16(2, 16, true);
        dv.setUint32(4, 24, true);
        dv.setUint32(8, node.lineNumber >>> 0, true);
        dv.setUint32(12, node.comment >>> 0, true);
        dv.setUint32(16, node.ns >>> 0, true);
        dv.setUint32(20, node.name >>> 0, true);
        return out;

      case 'cdata':
        out = new Uint8Array(28);
        dv = new DataView(out.buffer);
        dv.setUint16(0, CHUNK_XML_CDATA, true);
        dv.setUint16(2, 16, true);
        dv.setUint32(4, 28, true);
        dv.setUint32(8, node.lineNumber >>> 0, true);
        dv.setUint32(12, node.comment >>> 0, true);
        dv.setUint32(16, node.data >>> 0, true);
        dv.setUint16(20, node.valueSize, true);
        dv.setUint8(22, node.valueRes0);
        dv.setUint8(23, node.valueType);
        dv.setUint32(24, node.valueData >>> 0, true);
        return out;

      case 'elemStart':
        var attrCount = node.attributes.length;
        var size = 36 + attrCount * 20;
        out = new Uint8Array(size);
        dv = new DataView(out.buffer);
        dv.setUint16(0, CHUNK_XML_START_ELEMENT, true);
        dv.setUint16(2, 16, true);
        dv.setUint32(4, size, true);
        dv.setUint32(8, node.lineNumber >>> 0, true);
        dv.setUint32(12, node.comment >>> 0, true);
        dv.setUint32(16, node.ns >>> 0, true);
        dv.setUint32(20, node.name >>> 0, true);
        dv.setUint16(24, 20, true); // attributeStart
        dv.setUint16(26, 20, true); // attributeSize
        dv.setUint16(28, attrCount, true);
        dv.setUint16(30, node.idIndex || 0, true);
        dv.setUint16(32, node.classIndex || 0, true);
        dv.setUint16(34, node.styleIndex || 0, true);
        for (var a = 0; a < attrCount; a++) {
          var at = node.attributes[a];
          var ab = 36 + a * 20;
          dv.setUint32(ab, at.ns >>> 0, true);
          dv.setUint32(ab + 4, at.name >>> 0, true);
          dv.setUint32(ab + 8, at.rawValue >>> 0, true);
          dv.setUint16(ab + 12, at.valueSize, true);
          dv.setUint8(ab + 14, at.valueRes0);
          dv.setUint8(ab + 15, at.valueType);
          dv.setUint32(ab + 16, at.valueData >>> 0, true);
        }
        return out;

      case 'raw':
        return node.bytes;

      default:
        throw new Error('Unknown node type: ' + node.type);
    }
  }

  function serialize(doc) {
    var poolBytes = buildStringPool(doc.strings);
    var rmBytes = doc.resourceMapIds ? buildResourceMap(doc.resourceMapIds) : new Uint8Array(0);
    var nodeBytesList = doc.nodes.map(buildNode);

    var totalSize = 8 + poolBytes.length + rmBytes.length;
    for (var i = 0; i < nodeBytesList.length; i++) totalSize += nodeBytesList[i].length;

    var out = new Uint8Array(totalSize);
    var dv = new DataView(out.buffer);
    dv.setUint16(0, CHUNK_XML, true);
    dv.setUint16(2, 8, true);
    dv.setUint32(4, totalSize, true);

    var pos = 8;
    out.set(poolBytes, pos); pos += poolBytes.length;
    out.set(rmBytes, pos); pos += rmBytes.length;
    for (var i = 0; i < nodeBytesList.length; i++) {
      out.set(nodeBytesList[i], pos);
      pos += nodeBytesList[i].length;
    }
    return out;
  }

  function findString(strings, value) {
    for (var i = 0; i < strings.length; i++) {
      if (strings[i] === value) return i;
    }
    return -1;
  }

  function internString(doc, value) {
    var idx = findString(doc.strings, value);
    if (idx !== -1) return idx;
    doc.strings.push(value);
    return doc.strings.length - 1;
  }

  var AXML = {
    parse: parse,
    serialize: serialize,
    findString: findString,
    internString: internString,
    NO_INDEX: NO_INDEX
  };

  if (typeof module !== 'undefined' && module.exports) {
    module.exports = AXML;
  }
  global.AXML = AXML;
})(typeof window !== 'undefined' ? window : global);

File: permission-editor/app/src/main/java/site/ragdollp/permeditor/Axml.kt
package site.ragdollp.permeditor

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.Charset

/**
 * Minimal binary AndroidManifest.xml (AXML) reader/writer, ported from the
 * browser-side implementation in apk-permission-manager/axml.js. Implements
 * just enough of the AOSP ResourceTypes.h chunk format to round-trip a
 * manifest: string pool, optional resource map, and the flat
 * namespace/element/cdata node stream.
 *
 * `headerSize` for START_ELEMENT chunks is 16 (just the common node
 * header) - the attrExt fields (ns/name/attributeStart/...) are body, not
 * header, despite what a first read of the struct layout suggests. Getting
 * this wrong produces a document real AXML parsers (androguard, the
 * platform) reject even though a self-consistent reader/writer round-trips
 * it fine.
 */
object Axml {
    const val NO_INDEX = -1 // bit pattern 0xFFFFFFFF

    private const val CHUNK_STRING_POOL = 0x0001
    private const val CHUNK_XML = 0x0003
    private const val CHUNK_XML_START_NAMESPACE = 0x0100
    private const val CHUNK_XML_END_NAMESPACE = 0x0101
    private const val CHUNK_XML_START_ELEMENT = 0x0102
    private const val CHUNK_XML_END_ELEMENT = 0x0103
    private const val CHUNK_XML_CDATA = 0x0104
    private const val CHUNK_XML_RESOURCE_MAP = 0x0180

    data class Attribute(
        var ns: Int,
        var name: Int,
        var rawValue: Int,
        var valueSize: Int,
        var valueRes0: Int,
        var valueType: Int,
        var valueData: Int
    )

    sealed class Node {
        data class NsStart(var lineNumber: Int, var comment: Int, var prefix: Int, var uri: Int) : Node()
        data class NsEnd(var lineNumber: Int, var comment: Int, var prefix: Int, var uri: Int) : Node()
        data class ElemStart(
            var lineNumber: Int, var comment: Int, var ns: Int, var name: Int,
            var idIndex: Int, var classIndex: Int, var styleIndex: Int,
            var attributes: MutableList<Attribute>
        ) : Node()
        data class ElemEnd(var lineNumber: Int, var comment: Int, var ns: Int, var name: Int) : Node()
        data class Cdata(
            var lineNumber: Int, var comment: Int, var data: Int,
            var valueSize: Int, var valueRes0: Int, var valueType: Int, var valueData: Int
        ) : Node()
        data class Raw(var bytes: ByteArray) : Node()
    }

    class Doc(
        val strings: MutableList<String>,
        val resourceMapIds: IntArray?,
        val nodes: MutableList<Node>
    )

    fun findString(strings: List<String>, value: String): Int = strings.indexOf(value)

    fun internString(doc: Doc, value: String): Int {
        val idx = doc.strings.indexOf(value)
        if (idx != -1) return idx
        doc.strings.add(value)
        return doc.strings.size - 1
    }

    // ---------- parsing ----------

    private fun readLen16(bb: ByteBuffer, pos: Int): Pair<Int, Int> {
        val u0 = bb.getShort(pos).toInt() and 0xFFFF
        if (u0 and 0x8000 != 0) {
            val u1 = bb.getShort(pos + 2).toInt() and 0xFFFF
            return Pair(((u0 and 0x7FFF) shl 16) or u1, 4)
        }
        return Pair(u0, 2)
    }

    private fun readLen8(bytes: ByteArray, pos: Int): Pair<Int, Int> {
        val b0 = bytes[pos].toInt() and 0xFF
        if (b0 and 0x80 != 0) {
            val b1 = bytes[pos + 1].toInt() and 0xFF
            return Pair(((b0 and 0x7F) shl 8) or b1, 2)
        }
        return Pair(b0, 1)
    }

    private fun parseStringPool(bytes: ByteArray, bb: ByteBuffer, base: Int): Pair<Int, List<String>> {
        val chunkSize = bb.getInt(base + 4)
        val stringCount = bb.getInt(base + 8)
        val flags = bb.getInt(base + 16)
        val stringsStart = bb.getInt(base + 20)
        val utf8 = (flags and 0x100) != 0

        val offsets = IntArray(stringCount) { bb.getInt(base + 28 + it * 4) }
        val strings = ArrayList<String>(stringCount)
        for (i in 0 until stringCount) {
            val pos = base + stringsStart + offsets[i]
            if (utf8) {
                val (_, l1size) = readLen8(bytes, pos) // char length (unused)
                val (byteLen, l2size) = readLen8(bytes, pos + l1size)
                val dataPos = pos + l1size + l2size
                strings.add(String(bytes, dataPos, byteLen, Charsets.UTF_8))
            } else {
                val (charLen, lsize) = readLen16(bb, pos)
                strings.add(String(bytes, pos + lsize, charLen * 2, Charset.forName("UTF-16LE")))
            }
        }
        return Pair(chunkSize, strings)
    }

    fun parse(bytes: ByteArray): Doc {
        val bb = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
        val rootType = bb.getShort(0).toInt() and 0xFFFF
        require(rootType == CHUNK_XML) { "Not a binary XML document (root chunk type 0x${rootType.toString(16)})" }
        val rootSize = bb.getInt(4)

        var pos = 8
        require((bb.getShort(pos).toInt() and 0xFFFF) == CHUNK_STRING_POOL) { "Expected string pool chunk at $pos" }
        val (poolSize, strings) = parseStringPool(bytes, bb, pos)
        pos += poolSize

        var resourceMapIds: IntArray? = null
        if (pos + 8 <= bytes.size && (bb.getShort(pos).toInt() and 0xFFFF) == CHUNK_XML_RESOURCE_MAP) {
            val rmSize = bb.getInt(pos + 4)
            val n = (rmSize - 8) / 4
            resourceMapIds = IntArray(n) { bb.getInt(pos + 8 + it * 4) }
            pos += rmSize
        }

        val nodes = ArrayList<Node>()
        val end = rootSize
        while (pos < end) {
            val type = bb.getShort(pos).toInt() and 0xFFFF
            val headerSize = bb.getShort(pos + 2).toInt() and 0xFFFF
            val size = bb.getInt(pos + 4)
            val lineNumber = bb.getInt(pos + 8)
            val comment = bb.getInt(pos + 12)

            when (type) {
                CHUNK_XML_START_NAMESPACE, CHUNK_XML_END_NAMESPACE -> {
                    val prefix = bb.getInt(pos + 16)
                    val uri = bb.getInt(pos + 20)
                    nodes.add(
                        if (type == CHUNK_XML_START_NAMESPACE) Node.NsStart(lineNumber, comment, prefix, uri)
                        else Node.NsEnd(lineNumber, comment, prefix, uri)
                    )
                }
                CHUNK_XML_START_ELEMENT -> {
                    val ns = bb.getInt(pos + 16)
                    val name = bb.getInt(pos + 20)
                    val attributeStart = bb.getShort(pos + 24).toInt() and 0xFFFF
                    val attributeSize = bb.getShort(pos + 26).toInt() and 0xFFFF
                    val attributeCount = bb.getShort(pos + 28).toInt() and 0xFFFF
                    val idIndex = bb.getShort(pos + 30).toInt() and 0xFFFF
                    val classIndex = bb.getShort(pos + 32).toInt() and 0xFFFF
                    val styleIndex = bb.getShort(pos + 34).toInt() and 0xFFFF
                    val attrArrayBase = pos + 16 + attributeStart
                    val attributes = ArrayList<Attribute>(attributeCount)
                    for (a in 0 until attributeCount) {
                        val ab = attrArrayBase + a * attributeSize
                        attributes.add(
                            Attribute(
                                ns = bb.getInt(ab),
                                name = bb.getInt(ab + 4),
                                rawValue = bb.getInt(ab + 8),
                                valueSize = bb.getShort(ab + 12).toInt() and 0xFFFF,
                                valueRes0 = bb.get(ab + 14).toInt() and 0xFF,
                                valueType = bb.get(ab + 15).toInt() and 0xFF,
                                valueData = bb.getInt(ab + 16)
                            )
                        )
                    }
                    nodes.add(Node.ElemStart(lineNumber, comment, ns, name, idIndex, classIndex, styleIndex, attributes))
                }
                CHUNK_XML_END_ELEMENT -> {
                    nodes.add(Node.ElemEnd(lineNumber, comment, bb.getInt(pos + 16), bb.getInt(pos + 20)))
                }
                CHUNK_XML_CDATA -> {
                    val data = bb.getInt(pos + 16)
                    val vSize = bb.getShort(pos + 20).toInt() and 0xFFFF
                    val vRes0 = bb.get(pos + 22).toInt() and 0xFF
                    val vType = bb.get(pos + 23).toInt() and 0xFF
                    val vData = bb.getInt(pos + 24)
                    nodes.add(Node.Cdata(lineNumber, comment, data, vSize, vRes0, vType, vData))
                }
                else -> {
                    nodes.add(Node.Raw(bytes.copyOfRange(pos, pos + size)))
                }
            }
            pos += size
        }

        return Doc(strings.toMutableList(), resourceMapIds, nodes)
    }

    // ---------- serialization ----------

    private fun buildStringPool(strings: List<String>): ByteArray {
        val stringCount = strings.size
        val stringsStart = 28 + stringCount * 4

        data class Enc(val chars: String, val len: Int, val headBytes: Int, val byteLen: Int)
        val encoded = strings.map { s ->
            val len = s.length
            val headBytes = if (len >= 0x8000) 4 else 2
            Enc(s, len, headBytes, headBytes + len * 2 + 2)
        }
        var dataLen = encoded.sumOf { it.byteLen }
        while ((stringsStart + dataLen) % 4 != 0) dataLen++

        val chunkSize = stringsStart + dataLen
        val out = ByteArray(chunkSize)
        val bb = ByteBuffer.wrap(out).order(ByteOrder.LITTLE_ENDIAN)

        bb.putShort(0, CHUNK_STRING_POOL.toShort())
        bb.putShort(2, 28.toShort())
        bb.putInt(4, chunkSize)
        bb.putInt(8, stringCount)
        bb.putInt(12, 0) // styleCount
        bb.putInt(16, 0) // flags: UTF-16, not sorted
        bb.putInt(20, stringsStart)
        bb.putInt(24, 0) // stylesStart

        var offset = 0
        for (i in 0 until stringCount) {
            bb.putInt(28 + i * 4, offset)
            val e = encoded[i]
            val p = stringsStart + offset
            if (e.headBytes == 4) {
                bb.putShort(p, (0x8000 or (e.len ushr 16)).toShort())
                bb.putShort(p + 2, (e.len and 0xFFFF).toShort())
            } else {
                bb.putShort(p, e.len.toShort())
            }
            val charPos = p + e.headBytes
            for (c in 0 until e.len) {
                bb.putShort(charPos + c * 2, e.chars[c].code.toShort())
            }
            bb.putShort(charPos + e.len * 2, 0)
            offset += e.byteLen
        }
        return out
    }

    private fun buildResourceMap(ids: IntArray): ByteArray {
        val chunkSize = 8 + ids.size * 4
        val out = ByteArray(chunkSize)
        val bb = ByteBuffer.wrap(out).order(ByteOrder.LITTLE_ENDIAN)
        bb.putShort(0, CHUNK_XML_RESOURCE_MAP.toShort())
        bb.putShort(2, 8.toShort())
        bb.putInt(4, chunkSize)
        for (i in ids.indices) bb.putInt(8 + i * 4, ids[i])
        return out
    }

    private fun buildNode(node: Node): ByteArray = when (node) {
        is Node.NsStart -> buildNsLike(CHUNK_XML_START_NAMESPACE, node.lineNumber, node.comment, node.prefix, node.uri)
        is Node.NsEnd -> buildNsLike(CHUNK_XML_END_NAMESPACE, node.lineNumber, node.comment, node.prefix, node.uri)
        is Node.ElemEnd -> buildNsLike(CHUNK_XML_END_ELEMENT, node.lineNumber, node.comment, node.ns, node.name)
        is Node.Cdata -> {
            val out = ByteArray(28)
            val bb = ByteBuffer.wrap(out).order(ByteOrder.LITTLE_ENDIAN)
            bb.putShort(0, CHUNK_XML_CDATA.toShort())
            bb.putShort(2, 16.toShort())
            bb.putInt(4, 28)
            bb.putInt(8, node.lineNumber)
            bb.putInt(12, node.comment)
            bb.putInt(16, node.data)
            bb.putShort(20, node.valueSize.toShort())
            bb.put(22, node.valueRes0.toByte())
            bb.put(23, node.valueType.toByte())
            bb.putInt(24, node.valueData)
            out
        }
        is Node.ElemStart -> {
            val attrCount = node.attributes.size
            val size = 36 + attrCount * 20
            val out = ByteArray(size)
            val bb = ByteBuffer.wrap(out).order(ByteOrder.LITTLE_ENDIAN)
            bb.putShort(0, CHUNK_XML_START_ELEMENT.toShort())
            bb.putShort(2, 16.toShort()) // headerSize: node header only, ext is body
            bb.putInt(4, size)
            bb.putInt(8, node.lineNumber)
            bb.putInt(12, node.comment)
            bb.putInt(16, node.ns)
            bb.putInt(20, node.name)
            bb.putShort(24, 20.toShort()) // attributeStart
            bb.putShort(26, 20.toShort()) // attributeSize
            bb.putShort(28, attrCount.toShort())
            bb.putShort(30, node.idIndex.toShort())
            bb.putShort(32, node.classIndex.toShort())
            bb.putShort(34, node.styleIndex.toShort())
            for (a in 0 until attrCount) {
                val at = node.attributes[a]
                val ab = 36 + a * 20
                bb.putInt(ab, at.ns)
                bb.putInt(ab + 4, at.name)
                bb.putInt(ab + 8, at.rawValue)
                bb.putShort(ab + 12, at.valueSize.toShort())
                bb.put(ab + 14, at.valueRes0.toByte())
                bb.put(ab + 15, at.valueType.toByte())
                bb.putInt(ab + 16, at.valueData)
            }
            out
        }
        is Node.Raw -> node.bytes
    }

    private fun buildNsLike(type: Int, lineNumber: Int, comment: Int, a: Int, b: Int): ByteArray {
        val out = ByteArray(24)
        val bb = ByteBuffer.wrap(out).order(ByteOrder.LITTLE_ENDIAN)
        bb.putShort(0, type.toShort())
        bb.putShort(2, 16.toShort())
        bb.putInt(4, 24)
        bb.putInt(8, lineNumber)
        bb.putInt(12, comment)
        bb.putInt(16, a)
        bb.putInt(20, b)
        return out
    }

    fun serialize(doc: Doc): ByteArray {
        val poolBytes = buildStringPool(doc.strings)
        val rmBytes = doc.resourceMapIds?.let { buildResourceMap(it) } ?: ByteArray(0)
        val nodeBytesList = doc.nodes.map { buildNode(it) }

        var totalSize = 8 + poolBytes.size + rmBytes.size
        for (nb in nodeBytesList) totalSize += nb.size

        val out = ByteArrayOutputStream(totalSize)
        val header = ByteArray(8)
        val hb = ByteBuffer.wrap(header).order(ByteOrder.LITTLE_ENDIAN)
        hb.putShort(0, CHUNK_XML.toShort())
        hb.putShort(2, 8.toShort())
        hb.putInt(4, totalSize)
        out.write(header)
        out.write(poolBytes)
        out.write(rmBytes)
        for (nb in nodeBytesList) out.write(nb)
        return out.toByteArray()
    }
}
